pipeline/frontend/yaml/compiler/settings/params.go

```diff
@@ -0,0 +1,219 @@
+			return err
+		}
+	}
+	return nil
+}
+
+// format the environment variable key
+func sanitizeParamKey(k string) string {
+	return "PLUGIN_" + strings.ToUpper(
+		strings.ReplaceAll(strings.ReplaceAll(k, ".", "_"), "-", "_"))
+}
+
+// indicate if a data type can be turned into string without encoding as json
+func isComplex(t reflect.Kind) bool {
+	switch t {
+	case reflect.Bool,
+		reflect.String,
+		reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64,
+		reflect.Float32, reflect.Float64:
+		return false
+	default:
+		return true
+	}
+}
+
+// sanitizeParamValue returns the value of a setting as string prepared to be injected as environment variable
+func sanitizeParamValue(v interface{}, secrets map[string]string) (string, error) {
+	t := reflect.TypeOf(v)
+	vv := reflect.ValueOf(v)
+
+	switch t.Kind() {
+	case reflect.Bool:
+		return strconv.FormatBool(vv.Bool()), nil
+
+	case reflect.String:
+		return vv.String(), nil
+
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		return fmt.Sprintf("%v", vv.Int()), nil
+
+	case reflect.Float32, reflect.Float64:
+		return fmt.Sprintf("%v", vv.Float()), nil
+
+	case reflect.Map:
+		switch v := v.(type) {
+		// gopkg.in/yaml.v3 only emits this map interface
+		case map[string]interface{}:
+			// check if it's a secret and return value if it's the case
+			value, isSecret, err := injectSecret(v, secrets)
+			if err != nil {
+				return "", err
+			} else if isSecret {
+				return value, nil
+			}
+		default:
+			return "", fmt.Errorf("could not handle: %#v", v)
+		}
+
+		return handleComplex(vv.Interface(), secrets)
+
+	case reflect.Slice, reflect.Array:
+		if vv.Len() == 0 {
+			return "", nil
+		}
+
+		// if it's an interface unwrap and element check happen for each iteration later
+		if t.Elem().Kind() == reflect.Interface ||
+			// else check directly if element is not complex
+			!isComplex(t.Elem().Kind()) {
+			containsComplex := false
+			in := make([]string, vv.Len())
+
+			for i := 0; i < vv.Len(); i++ {
+				v := vv.Index(i).Interface()
+
+				// ensure each element is not complex
+				if isComplex(reflect.TypeOf(v).Kind()) {
+					containsComplex = true
+					break
+				}
+
+				var err error
+				if in[i], err = sanitizeParamValue(v, secrets); err != nil {
+					return "", err
+				}
+			}
+
+			if !containsComplex {
+				return strings.Join(in, ","), nil
+			}
+		}
+	}
+
+	// handle all elements which are not primitives, string-maps containing secrets or arrays
+	return handleComplex(vv.Interface(), secrets)
+}
+
+// handleComplex uses yaml2json to get json strings as values for environment variables
+func handleComplex(v interface{}, secrets map[string]string) (string, error) {
+	v, err := injectSecretRecursive(v, secrets)
+	if err != nil {
+		return "", err
+	}
+
+	out, err := yaml.Marshal(v)
+	if err != nil {
+		return "", err
+	}
+	out, err = yaml2json.Convert(out)
+	if err != nil {
+		return "", err
+	}
+	return string(out), nil
```

Comment:
because json key words would be upercase...
```diff
- (string) (len=14) "PLUGIN_COMPLEX": (string) (len=33) "[{\"name\":\"Jack\"},{\"name\":\"Jill\"}]",
- (string) (len=15) "PLUGIN_COMPLEX2": (string) (len=15) "{\"name\":\"Jack\"}",
+ (string) (len=14) "PLUGIN_COMPLEX": (string) (len=33) "[{\"Name\":\"Jack\"},{\"Name\":\"Jill\"}]",
+ (string) (len=15) "PLUGIN_COMPLEX2": (string) (len=15) "{\"Name\":\"Jack\"}",
```

Reply:
if we want to throw out this conversion fine ... but let this bugfix be merged first